oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,393 @@
+            result
+        }
+
+        Quantile {
+            marker_hghts: sub_arrays(&self.marker_hghts, &other.marker_hghts),
+            marker_pos: sub_arrays(&self.marker_pos, &other.marker_pos),
+            desired_marker_pos: sub_arrays(
+                &self.desired_marker_pos,
+                &other.desired_marker_pos,
+            ),
+            desired_marker_incrs: sub_arrays(
+                &self.desired_marker_incrs,
+                &other.desired_marker_incrs,
+            ),
+        }
+    }
+
+    /// Parabolic prediction for marker height.
+    fn parabolic(&self, i: usize, d: f64) -> f64 {
+        let term1 =
+            d / (self.marker_pos[i + 1] - self.marker_pos[i - 1]) as f64;
+        let term2 = ((self.marker_pos[i] - self.marker_pos[i - 1]) as f64 + d)
+            * (self.marker_hghts[i + 1] - self.marker_hghts[i])
+            / (self.marker_pos[i + 1] - self.marker_pos[i]) as f64;
+        let term3 = ((self.marker_pos[i + 1] - self.marker_pos[i]) as f64 - d)
+            * (self.marker_hghts[i] - self.marker_hghts[i - 1])
+            / (self.marker_pos[i] - self.marker_pos[i - 1]) as f64;
+
+        self.marker_hghts[i] + term1 * (term2 + term3)
+    }
+
+    /// Linear prediction for marker height.
+    fn linear(&self, i: usize, d: f64) -> f64 {
+        let index = if d < 0. { i - 1 } else { i + 1 };
+        self.marker_hghts[i]
+            + d * (self.marker_hghts[index] - self.marker_hghts[i])
+                / (self.marker_pos[index] - self.marker_pos[i]) as f64
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use oximeter::test_util::assert_almost_eq;
+
+    #[test]
+    fn test_float_observations() {
+        let observations = [
+            0.02, 0.5, 0.74, 3.39, 0.83, 22.37, 10.15, 15.43, 38.62, 15.92,
+            34.60, 10.28, 1.47, 0.40, 0.05, 11.39, 0.27, 0.42, 0.09, 11.37,
+        ];
+        let mut q = Quantile::p50();
+        for &o in observations.iter() {
+            q.append(o).unwrap();
+        }
+        assert_eq!(q.marker_pos, [1, 6, 10, 16, 20]);
+        assert_eq!(q.desired_marker_pos, [1., 5.75, 10.50, 15.25, 20.0]);
+        assert_eq!(q.len(), 20);
+        assert_eq!(q.p(), 50);
+        assert_almost_eq!(q.estimate().unwrap(), 4.2462394088036435, 2e-15);
+    }
+
+    #[test]
+    fn test_integer_observations() {
+        let observations = 1..=100;
+        let mut q = Quantile::new(30).unwrap();
+        for o in observations {
+            q.append(o).unwrap();
+        }
+
+        assert_eq!(q.marker_pos, [1, 15, 30, 65, 100]);
+        assert_eq!(
+            q.desired_marker_pos,
+            [
+                1.0,
+                15.850000000000026,
+                30.70000000000005,
+                65.34999999999992,
+                100.0
+            ]
+        );
+        assert_eq!(q.len(), 100);
+        assert_eq!(q.p(), 30);
+        assert_eq!(q.estimate().unwrap(), 30.0);
+    }
+
+    #[test]
+    fn test_p50() {
```

Comment:
The following tests are the same internals, could we factor out the guts to something like:

```rust
test_quantile_impl(p: u64) {
    let mut q = Quantile::new(p).unwrap();
    for o in 1..=100 {
        q.append(o).unwrap();
    }
    assert_eq!(q.p(), p);
    assert_eq!(q.estimate().unwrap(), f64::try_from(p).unwrap());
}
```

Reply:
yep, just did this today actually :).